README.md
# Snakebite-Metagenomics

## Background and Features
A Snakemake - Metagenomics pipeline that covers the main aspects of shotgun
metagenomics.

The pipeline consists of different modules, namely

* QC
* Metagenome assembly
  * Full assembly, using all samples
  * Co-assembly subsets
* Gene prediction
* Read alignments
* Binning
* Taxonomic annotation

## Requirements
All required tools are defined as singularity / docker container that are
downloaded when needed. Further, the pre-configured resource allocations are
aimed for a slurm queueing system, other options are naturally possible, but
require then more user interaction. 

The pipeline works out-of-the-box, if the system has installed

* Docker / singularity /apptainer
* slurm

## Installation
The pipeline can be installed by cloning the corresponding GitHub repository

```
git clone https://github.com/fischuu/Snakebite-Metagenomics.git
```

## Preparation of the pipeline
There are three essential configuration files that needs to be adjusted

* `Snakebite-Metagenomics_config.yaml`
* `Snakebite-Metagenomics_server_config.yaml`
* `run_Snakebite-Metagenomics.sh`

The pipeline is shipped with an example configuration that can be adjusted to the
local needs.

In additon to that, a sample sheet needs to be created, an example for it is
also shipped with the pipeline

* `sampleSheet.tsv`

## Starting the pipeline
When all configuration files are filled, you can start the pipeline by running
the start script like this

```
bash run_Snakebite-Metagenomics.sh
```

## More details
For more details, please visit the project wiki page:

https://github.com/fischuu/Snakebite-Metagenomics/wiki

It contains details to the runtime paramters, explains how to populate the above
mentioned files and also contains convenience script to create the samplesheet
programmatically.

Further, the pipeline has several entry points so that it does not need to run
as a whole.

CHANGELOG.md
# CHANGELOG

## PLANNED UPDATES FOR FUTURE VERSIONS

## UPDATES

### Development version 0.5.*:
--------------------------------------------------------------------------------
9: Keep the concatenated RAW reads and create md5 sums for them

8: Typo in the DECONTAMINATION FastQC rule fixed

7: Added the QC part to the final report

6: Added the basicStats section to the final report

5: Added the general workflow section to the final report

4: Added the helpFunctions section to the final report

3: Now is the default server configuration expected to be in the project folder

2: Added the template to prepare the summary statistics

1: Added the final report template


### Release version 0.4:
--------------------------------------------------------------------------------

### Development version 0.3.*:
--------------------------------------------------------------------------------
29: Added sed command to remove comment lines from the prodigal gene prediction

28: Added the k-min option for megahit to avoid crashes in case of deep sequenced genomes

27: Adjusted the output paths for the MAG module

26: Fixed a bug in the EGGNOG annotation step, that prevented the annotation to happen...

25: Added the concoct steps for the co-assemblies

24: Some concoct intermediate files are set to temp now

23: Concoct module  for the full assembly finalised

22: Added the --cluster-cancel option to the start script

21: Added the filtered assemblies to the QUAST report

20: Added a length filter for the metagenome assemblies

19: Constrained the cagroup wildcard

18: Report reads per contig for full and coas

17: Added alignment statistics to the quantification

16: Quantification step added

15: Eggnog annotation added

14: Prodigal gene prediction added

13: Samples can now be used in several co-assemblies, separated by comma in samplesheet file

12: Server resource allocation names adjusted

11: Bowtie2 indices are now forced to be large to avoid index naming issues

10: Added rules to align reads against the various assemblies (full and co-assemblies)

9: Metaquast added to the pipeline

8: Group-wise co-assemblies are now also generated

7: Added the QC (FastQC and MultiQC) for the Raw samples and concatenated as well as trimmed ones

6: Example start and config files copy+paste error fixed

5: More adjustments to the default resource allocations (extract_fasta_concoct_1k and extract_fasta_concoct_2k)

4: MEGAHIT_temp folder is now temporary only, to save tons of disc space

3: Adjusted the defaults resources for concoct parts 1k and 2k

2: Critical bugfix, 2k concoct calculations were done with 1k output

1: Inital submission

### Release version 0.2, includes all changes from dev version 0.1.*
--------------------------------------------------------------------------------

### Development version 0.1.*:
--------------------------------------------------------------------------------
16: Default resource allocations adjusted for avoid crashes

15: rule extract_fasta_concoct for 1k and 2k added

14: rule run_concoct added for 1k and 2k

13: rule cut_filtered_contigs_concoct added

12: Step5-MAGs.smk ruleset started

11: rule quantify_predictedGenes_featureCounts added

10: rule filter_length_megahit added

9: Eggnot orthology rule added

8: Eggnog gene annotation added

7: Execution script adjusted to bind local nvme discs

6: Prodigal output is now cut in chunks for parallel processing

5: Eggnog databases are now downloaded automatically

4: Bugfix in bowtie2 index creation step fixed

3: create_index_bowtie2 write now its output into a logfile

2: Added the decontamination step

1: Inital submission


scripts/R/Final-report.R
last_n_char <- function(x, n=1){
  substr(x, nchar(x) - n + 1, nchar(x))
}

if(last_n_char(pipelineFolder)!="/") pipelineFolder <- paste0(pipelineFolder, "/")

createRMD.command <- paste0("cat ",pipelineFolder,"scripts/Rmd/final-header.Rmd ",
                                   pipelineFolder,"scripts/Rmd/helpFunctions.Rmd ",
                                   pipelineFolder,"scripts/Rmd/generalWorkflow.Rmd ",
                                   pipelineFolder,"scripts/Rmd/basicStats.Rmd ",
                                   pipelineFolder,"scripts/Rmd/QC.Rmd ",
                                   "> ",projFolder,"/finalReport.Rmd")


cat(createRMD.command, "\n")

system(createRMD.command)

rmarkdown::render(file.path(projFolder,"finalReport.Rmd"), output_file=file.path(projFolder,"finalReport.html"))

run_Metagenome-Pipeline.sh
# This conda module is just to make snakemake available
module load Snakemake/7.17.1

pipelineFolder="/users/XXX/git/Pipeline-GBS"
projectFolder="/scratch/project_XXX/"

# Create the rulegraph
snakemake -s $pipelineFolder/Snakefile-Pipeline-Metagenome.smk \
          --configfile $projectFolder/Pipeline-Metagenome_config.yaml \
          --rulegraph | dot -T png > $projectFolder/workflow.png

snakemake -s $pipelineFolder/Snakefile-Pipeline-Metagenome.smk \
          -j 300 \
          --use-singularity \
          --singularity-args "-B /scratch:/scratch,/run:/run" \
          --configfile $projectFolder/Pipeline-Metagenome_config.yaml \
          --latency-wait 60 \
          --cluster-config $pipelineFolder/Pipeline-Metagenome_server_config.yaml \
          --cluster "sbatch -t {cluster.time} --account={cluster.account} --gres=nvme:{cluster.nvme} --job-name={cluster.job-name} --tasks-per-node={cluster.ntasks} --cpus-per-task={cluster.cpus-per-task} --mem-per-cpu={cluster.mem-per-cpu} -p {cluster.partition} -D {cluster.working-directory} --parsable" \
          --cluster-cancel scancel $@


scripts/prepareSummaryData.R
# This is a development script to summarise the data it will be later generalised and added to the pipeline

projFolder <- "/scratch/project_2006940/MASTER"
coas <- 2

library("GenomicTools")
library("data.table")

filename <- file.path(projFolder, "PRODIGAL", paste0("final.contigs_group_", coas), paste0("final.contigs_group_",coas,".prodigal_plain.gtf"))

prodGFF <- importGFF(filename, level="CDS")

# Now import the FeatureCounts quantification

fc.files <- list.files(file.path(projFolder, "QUANTIFICATION", "PRODIGAL_FC", paste0("final.contigs_group_", coas)), pattern="*.txt$")

# This is now for me to speed up things, remove it later for production
fc.files <- fc.files[grep("FEC", fc.files)]

tmp <- importFeatureCounts(file.path(projFolder, "QUANTIFICATION", "PRODIGAL_FC", paste0("final.contigs_group_", coas), fc.files[1]))
tmpExp <- tmp$expValues
colnames(tmpExp) <- gsub("X.scratch.project_2006940.MASTER.BAM.final.contigs_coas2.LUKE_", "", colnames(tmpExp))
colnames(tmpExp) <- gsub("_mega.bam", "", colnames(tmpExp))

fcExp <- matrix(0, ncol=length(fc.files)+2, nrow=nrow(tmpExp))
fcExp <- as.data.table(fcExp)
rownames(fcExp) <- tmpExp[,1]
colnames(fcExp)[1] <- "Geneid"
colnames(fcExp)[2] <- "Eggnog_id"
colnames(fcExp)[3] <- colnames(tmpExp)[2]
fcExp[,1] <- tmp$geneInfo$Geneid
fcExp[,2] <- paste0(tmp$geneInfo$Chr,"_",sapply(strsplit(tmp$geneInfo$Geneid, "_"),"[",2))
fcExp[,3] <- tmpExp[,2]

for(i in 2:length(fc.files)){
  tmp <- importFeatureCounts(file.path(projFolder, "QUANTIFICATION", "PRODIGAL_FC", paste0("final.contigs_group_", coas), fc.files[i]))
  tmpExp <- tmp$expValues
  colnames(tmpExp) <- gsub("X.scratch.project_2006940.MASTER.BAM.final.contigs_coas2.LUKE_", "", colnames(tmpExp))
  colnames(tmpExp) <- gsub("_mega.bam", "", colnames(tmpExp))

  colnames(fcExp)[i+2] <- colnames(tmpExp)[2]
  fcExp[,i+2] <- tmpExp[,2]
  
  
    cat("Sample",i,"/", length(fc.files),"processed -",date(),"\n")
}

fwrite(fcExp, file.path(projFolder,"fcExp_coas2.csv"))


run_Snakebite-Metagenomics.sh
module load snakemake

pipelineFolder="/users/fischerd/git/Snakebite-Metagenomics"
projectFolder="/scratch/project_2001746/Metagenomics_Example"

export APPTAINER_TMPDIR="/scratch/project_2001746/tmp"
export APPTAINER_CACHEDIR="/scratch/project_2001746/tmp"

mkdir -p $APPTAINER_TMPDIR

# Create the rulegraph
snakemake -s $pipelineFolder/Snakebite-Metagenomics.smk \
          --configfile $projectFolder/Snakebite-Metagenomics_config.yaml \
          --rulegraph | dot -T png > $projectFolder/workflow.png

snakemake -s $pipelineFolder/Snakebite-Metagenomics.smk \
          -j 300 \
          --use-singularity \
          --singularity-args "-B /scratch:/scratch,/run:/run,/users:/users" \
          --configfile $projectFolder/Snakebite-Metagenomics_config.yaml \
          --latency-wait 60 \
          --cluster-config $projectFolder/Snakebite-Metagenomics_server_config.yaml \
          --cluster "sbatch -t {cluster.time} --account={cluster.account} --gres=nvme:{cluster.nvme} --job-name={cluster.job-name} --tasks-per-node={cluster.ntasks} --cpus-per-task={cluster.cpus-per-task} --mem-per-cpu={cluster.mem-per-cpu} -p {cluster.partition} -D {cluster.working-directory} --parsable" \
          --cluster-cancel scancel $@
